From efd777d188d7c40b7d920a87728b8e7c60b23f20 Mon Sep 17 00:00:00 2001
From: Adrian Vovk <avovk@gnome.org>
Date: Wed, 17 Jun 2026 22:01:35 -0400
Subject: [PATCH] Add support for systemd-homed

Enable homed unconditionally: switch gnome-initial-setup to Adrian
Vovk's homed fork for proper user creation; drop malcontent from
gnome-control-center, gnome-initial-setup, and gnome-software;
disable parental controls flags; enable -Dcreate_homed=true in
accountsservice; configure the live generator to set
DefaultStorage=subvolume for homed home directories.

Port of gnome-build-meta!2681.

Upstream-Status: Submitted https://gitlab.gnome.org/GNOME/gnome-build-meta/-/merge_requests/2681
Exit condition: Drop after gnome-build-meta gnome-50 merges MR !2681
---
 elements/core-deps/accountsservice.bst | 4 ++++
 elements/core/gnome-control-center.bst | 2 --
 elements/core/gnome-initial-setup.bst  | 9 ++++-----
 elements/core/gnome-software.bst       | 2 +-
 files/gnomeos/live/generator.c         | 5 +++++
 5 files changed, 21 insertions(+), 8 deletions(-)

diff --git a/elements/core-deps/accountsservice.bst b/elements/core-deps/accountsservice.bst
index dabcfa8..73f280f 100644
--- a/elements/core-deps/accountsservice.bst
+++ b/elements/core-deps/accountsservice.bst
@@ -18,3 +18,7 @@ depends:
 - freedesktop-sdk.bst:components/dbus.bst
 - freedesktop-sdk.bst:components/polkit.bst
 - freedesktop-sdk.bst:public-stacks/runtime-minimal.bst
+
+variables:
+  meson-local: >-
+    -Dcreate_homed=true
diff --git a/elements/core/gnome-control-center.bst b/elements/core/gnome-control-center.bst
index 3e59f75..babd028 100644
--- a/elements/core/gnome-control-center.bst
+++ b/elements/core/gnome-control-center.bst
@@ -39,7 +39,6 @@ depends:
 - core-deps/ibus-daemon.bst
 - core-deps/libgtop.bst
 - core-deps/libnma.bst
-- core-deps/malcontent.bst
 - core-deps/samba.bst
 - core-deps/system-config-printer.bst
 - core-deps/udisks2.bst
@@ -64,4 +63,3 @@ depends:
 variables:
   meson-local: >-
     -Ddocumentation=true
-    -Dmalcontent=true
diff --git a/elements/core/gnome-initial-setup.bst b/elements/core/gnome-initial-setup.bst
index d207fb4..98f1d9c 100644
--- a/elements/core/gnome-initial-setup.bst
+++ b/elements/core/gnome-initial-setup.bst
@@ -2,9 +2,9 @@ kind: meson
 
 sources:
 - kind: git_repo
-  url: gnome:gnome-initial-setup.git
-  track: master
-  ref: 50.0-66-g77be510de42c5b14c0967b6f61609990d0c2505b
+  url: gnome_gitlab:AdrianVovk/gnome-initial-setup.git
+  track: homed
+  ref: 50.0-10-g952d28e3e3aefe67bc0dd2d37a0be964d2b69d66
 
 build-depends:
 - freedesktop-sdk.bst:public-stacks/buildsystem-meson.bst
@@ -18,7 +18,6 @@ depends:
 - core-deps/libgweather.bst
 - core-deps/libnma.bst
 - core-deps/librest.bst
-- core-deps/malcontent.bst
 - core-deps/upower.bst
 - core/gdm.bst
 - core/gnome-desktop.bst
@@ -34,4 +33,4 @@ depends:
 
 variables:
   meson-local: >-
-    -Dparental_controls=enabled
+    -Dparental_controls=disabled
diff --git a/elements/core/gnome-software.bst b/elements/core/gnome-software.bst
index 29e7c47..5c46b2b 100644
--- a/elements/core/gnome-software.bst
+++ b/elements/core/gnome-software.bst
@@ -20,7 +20,6 @@ depends:
 - core-deps/fwupd.bst
 - core-deps/glib-testing.bst
 - core-deps/gnome-online-accounts.bst
-- core-deps/malcontent.bst
 - core-deps/snapd-glib.bst
 - core/gnome-desktop.bst
 - sdk/adwaita-icon-theme.bst
@@ -49,3 +48,4 @@ variables:
     -Dpackagekit=false
     -Dsnap=true
     -Dsystemd-sysupdate=true
+    -Dmalcontent=false
diff --git a/files/gnomeos/live/generator.c b/files/gnomeos/live/generator.c
index 8fe5b0f..bf92377 100644
--- a/files/gnomeos/live/generator.c
+++ b/files/gnomeos/live/generator.c
@@ -154,6 +154,18 @@ static bool handle_live(const char *generator_path, bool in_initrd) {
                 fprintf(file, "[zram1]\nmount-point=/\nfs-type=btrfs\n");
                 fclosep(&file);
 
+                if (mkdir("/run/systemd/homed.conf.d", 0755) < 0 && errno != EEXIST) {
+                        kmsg_printf(LOG_WARNING, "Failed to create homed.conf.d directory: %m\n");
+                } else {
+                        file = fopen("/run/systemd/homed.conf.d/00-force-subvolume.conf", "w");
+                        if (file) {
+                                fprintf(file, "[Home]\nDefaultStorage=subvolume\n");
+                                fclosep(&file);
+                        } else {
+                                kmsg_printf(LOG_WARNING, "Failed to create homed configuration: %m\n");
+                        }
+                }
+
                 if (!mark_live())
                         return false;
 
-- 
2.54.0

